File: USACO Training/3.1/Score Inflation Java/src/inflate.java
/*
ID: cormac.1
LANG: JAVA
TASK: inflate
 */

import java.io.*;
import java.util.*;
import java.lang.*;

public class inflate {
    public static void main(String[] args) throws IOException {
        BufferedReader fin = new BufferedReader(new FileReader("inflate.in"));
        PrintWriter fout = new PrintWriter(new BufferedWriter(new FileWriter("inflate.out")));

        StringTokenizer s = new StringTokenizer(fin.readLine());

        int M = Integer.parseInt(s.nextToken());
        int N = Integer.parseInt(s.nextToken());

        int[] points = new int[N];
        int[] minutes = new int[N];

        for (int i = 0; i < N; ++i) {
            s = new StringTokenizer(fin.readLine());

            points[i] = Integer.parseInt(s.nextToken());
            minutes[i] = Integer.parseInt(s.nextToken());
        }

        int[] dp = new int[M+1];
        dp[0] = 0;

        for (int t = 1; t <= M; t++) {
            dp[t] = dp[t-1];
            for (int i = 0; i < N; ++i) {
                if (minutes[i] > t) continue;
                dp[t] = Math.max(dp[t], dp[t - minutes[i]] + points[i]);
            }
        }

        fout.println(dp[M]);
        fout.close();
    }
}

File: USACO Training/3.1/Score Inflation/inflate.py
"""
TASK: inflate
ID: cormac.1
LANG: PYTHON3
"""

with open("inflate.in") as file:
    M, N = map(int, file.readline().split())

    # (points, minutes)
    classes = [tuple(map(int, file.readline().split())) for i in range(N)]

classes.sort(key=lambda n: n[1])

dp = [0 for i in range(M+1)]

for t in range(1, M+1):
    dp[t] = dp[t-1]
    for cls in classes:
        if cls[1] <= t:
            dp[t] = max(dp[t], dp[t-cls[1]] + cls[0])
        else:
            break

with open("inflate.out", "w") as file:
    file.write(str(dp[M]) + "\n")
        
